### ISSUE TITLE
API migration timeline. Continuation of side-discussion in #674

### ISSUE THREAD

**Issue by kbevers** [OSGeo/PROJ#686]
I asked an off-topic question in https://github.com/OSGeo/proj.4/pull/674#issuecomment-347221630 regarding the plan ahed API wise. Continuing the discussion here as to not clutter the existing thread.

My question:

> Do you mind outlining such a plan? At the moment we are all working in sort of the same direction, but my perception is that there is no common understanding of what exactly that direction is. It would be nice to have that settled once and for all. Coming from you, the project maintainer, it can be regarded as the "authorative plan" I guess.

to which @hobu answered:

> Well, I'm the self-appointed maintainer, and I'm happy to hand off the hat to people actually doing all the work 😄 I've been trying to stay back and let you guys evolve things without worrying too much about the project maintenance implications.
> 
> In my opinion, the 5.0.0 release cannot break projects that are using the existing API, except to force them to recompile. I think it is just fine to annoyingly warn them the API they're using is deprecated, and in one or two releases, require them to #define I_KNOW_THIS_API_IS_OLD to continue to use the old API. We can't force a migration by breaking compilation of existing usage that's out there in one release though, or we'll end up splitting the project into two, and the packagers will choose the old one and start patching because of all of the existing software.
> 
> ![image](https://user-images.githubusercontent.com/87751/33276224-8b2d0fb2-d35a-11e7-8235-f578adb189e3.png)
> 
> Yes I know this ends up being a little bit like XKCD, except for APIs instead of standards, but we don't have much choice. The library is 30 years old, and we don't get to break people for our own satisfaction -- no matter how right, correct, fast, or convenient our new reality might be.

**Comment by kbevers** [comment]
> Well, I'm the self-appointed maintainer, and I'm happy to hand off the hat to people actually doing all the work 😄 

If you want to step down I am sure we can find someone to take over the reins (if no one else volunteers I'll consider it). If that is the case, Ideally that would be after 5.0.0 is out the door. 

> I've been trying to stay back and let you guys evolve things without worrying too much about the project maintenance implications.

And that's been great so far. As we are getting close to a new release, I think now is the time to worry about those implications.

> In my opinion, the 5.0.0 release cannot break projects that are using the existing API, except to force them to recompile. I think it is just fine to annoyingly warn them the API they're using is deprecated, and in one or two releases, require them to #define I_KNOW_THIS_API_IS_OLD to continue to use the old API. We can't force a migration by breaking compilation of existing usage that's out there in one release though, or we'll end up splitting the project into two, and the packagers will choose the old one and start patching because of all of the existing software.

I agree.

> Yes I know this ends up being a little bit like XKCD, except for APIs instead of standards, but we don't have much choice. The library is 30 years old, and we don't get to break people for our own satisfaction -- no matter how right, correct, fast, or convenient our new reality might be.

I also agree with this. It is a pain in the arse having the older stuff to take care of for a number of years, but that's the way it is. Having a schedule for when the old stuff will be removed would however make it more bearable for me. Something like remove `projects.h` from the public API with version 6.0.0 and `proj_api.h` with 7.0.0 for instance (and maybe promising a life span of a few years for each of the next 2 major versions release). Not only would that be nice from a development perspective, but it would also give users of the library some sort of time frame in which they have to migrate to the new API.

**Comment by hobu** [comment]
consider this as a notional attempt

# 5.0.0 -- 01 FEB 2018
* [ ] New API
* [ ] New transformation capabilities
* [ ] Significant documentation updates
* [ ] API migration guide
* [ ] API deprecation warning for `projects.h` and `proj_api.h`
* [ ] Put in release notes that WinCE will not be supported from 5.1.0 and onwards.

## 5.1.0 -- 01 MAY 2018
* [ ] New API revisions in response to actual use
* [ ] Bug fixes
* [ ] Security patches
* [ ] Packager upstream additions
* [ ] Remove WinCE support (#582)

## 5.2.0 -- 01 AUG 2018
* [ ] New API revisions and refinements
* [ ] Bug fixes
* [ ] Security patches
* [ ] Packager upstream additions

# 6.0.0 -- 01 FEB 2019
* [ ] Removal of `projects.h`
* [ ] `#define I_KNOW_THIS_API_IS_OLD` (or something named more friendly) to use  `proj_api.h` without compilation error
* [ ] New API freeze
* [ ] Bug fixes
* [ ] Security patches

## 6.1.0 -- 01 AUG 2019
* [ ] Bug fixes
* [ ] Security patches


# 7.0.0 -- 01 FEB 2020
* [ ] Removal of  `proj_api.h` 

# Identification of API impact
> it would also give users of the library some sort of time frame in which they have to migrate to the new API.

Some things will never migrate, and we have to be ok with that. @sebastic's tools for identifying packages that are using old APIs will be helpful, and I'm sure some of the PROJ developers can go to those projects with patches that implement the migration(s). It should be straightforward for most projects, it's just the effort required to do so and test.

**Comment by kbevers** [comment]
Thanks, Howard. I think this is a great plan. I am sure some will say that it is too fast paced, but that is likely to be the case however long we wait to remove the old API's.

I'll make sure to do my part in helping depending projects migrate so we get as many as possible on board from the start. Greg Troxel suggested that we open issues on as many affected projects issue trackers as possible. I think that is a good idea and I'll give it a go once we a new release in the wild.

**Comment by sebastic** [comment]
I don't have any specific tools for identifying packages that are using the old APIs, just [ben](http://ben.debian.net/) to generate a list of packages dependening on libproj (which `apt-cache rdepends libproj12` can also do) and [codesearch](https://codesearch.debian.net/) to grep through all package sources to find which headers they include.

**Comment by QuLogic** [comment]
I suggest adding `I_KNOW_THIS_API_IS_OLD` (or whatever the name) be added in 5.0.0 (but defaulting off) so that people that want to be sure they're only using new API can set the flag and do so as early as possible.

Of course, that's only required if the deprecated and non-deprecated API are not confined to separate headers. If they are totally distinct, then it's as simple as not including one or the other, and this doesn't need to be done.

**Comment by kbevers** [comment]
> Of course, that's only required if the deprecated and non-deprecated API are not confined to separate headers. If they are totally distinct, then it's as simple as not including one or the other, and this doesn't need to be done.

The new API (`proj.h`) is orthogonal to the old stuff (`projects.h` + `proj_api.h`), so that shouldn't be necessary.

**Comment by hobu** [comment]
> The new API (proj.h) is orthogonal to the old stuff (`projects.h` + `proj_api.h`), so that shouldn't be necessary.

Is mixing to be disallowed entirely? Would it be possible and/or practical to disallow mixing `proj.h` and (`projects.h`/`proj_api.h`) altogether?

**Comment by kbevers** [comment]
> Is mixing to be disallowed entirely? Would it be possible and/or practical to disallow mixing proj.h and (projects.h/proj_api.h) altogether?

Just took a closer look. At the moment you are allowed to mix `projects.h`/`proj_api.h` with `proj.h` as long as `proj.h` is included first. It is likely that this is allowed because it is needed internally. Not entirely sure about that though, @busstoptaktik knows the answer to this I am sure. I think it would be preferable if they are not allowed to be mixed in the public API.

**Comment by busstoptaktik** [comment]
I believe that the non-API roles now played by proj_api.h and projects.h should be taken over by proj_internal.h, and the API roles by proj.h.

The dual header structure enables information hiding.  For example, in proj.h, the typedef for PJ is based on a forward declaration of the struct PJconsts:

    struct PJconsts;
    typedef struct PJconsts PJ;         /* the PJ object herself */

The full definition of `struct PJconsts` is given in `projects.h`, so when including just `proj.h` it is impossible to accidentally change a field in a `PJ`.

We *could* start a migration, where we allow `proj_internal.h` to grow quite a bit, and duplicate whatever is needed (including the PJconsts definition), from `projects.h`.

But it will be a hassle to make sure that the duplicated material stays compatible. On the other hand, it would give us a chance to begin to eliminate the internal dependency on `projects.h`, which would be awesome.

So in summary: Yes, currently we need a level of compatibility between `proj.h` and `projects.h` (which in turn includes `proj_api.h`), for library-internal use. This, of course, also means that we could disallow concurrent use unless the `PJ_LIB__` symbol is defined.

We have come far since the clean up efforts started, so I think it *may* be time to begin on the *eliminate internal dependency on `projects.h`* part as well.

**Comment by kbevers** [comment]
> So in summary: Yes, currently we need a level of compatibility between proj.h and projects.h (which in turn includes proj_api.h), for library-internal use. This, of course, also means that we could disallow concurrent use unless the PJ_LIB__ symbol is defined.

I think this is the way to go. I fear that if we allow mixing `proj.h` with the old header files people will do it. If we dissallow it, we will hopefully get reports from users that are missing functionality in the new API and we can act accordingly.

> We have come far since the clean up efforts started, so I think it may be time to begin on the eliminate internal dependency on projects.h part as well.

This would be great, but to me this sounds like something that should wait until after 5.0.0. Perhaps even till after 5.2.0 (if it requires duplicating code at least).

**Comment by kbevers** [comment]
On the mailing list we've previously discussed getting rid of the nmake build system (CMake can be used instead) and the WinCE bindings (no one apparently uses them). I believe we came to the agreement that both needs to go, it's just a matter of when. The easy way is to remove both before 5.0.0 is out the door (not sure how big the impact will be). Or we can wait until one of the minor releases of v.5 and make sure to mention it in the release notes. In that case they should be a part of the plan above.

Thoughts?

**Comment by hobu** [comment]
There's some stuff that would need to be done to bring the nmake stuff in line with CMake. 

# NMake migration to CMake checklist
* [ ] DLL version naming
* [ ] Manifest files
* [ ] `proj.def` handling
* [ ] `nad2bin.exe` execution
* [ ] `nmake.opt` catch up

**Comment by kbevers** [comment]
I added a few points regarding WinCE support to the migration plan above.

The  VB6 stuff in `com/` stuff seems to be in similar disrepair. Maybe we should consider getting rid of that too?

**Comment by mloskot** [comment]
I think both, WinCE and VB6, could really be dropped as soon as 5.0.0 - a thick line, if anyone asks for it, "Sure, use 4.x, luke"

**Comment by kbevers** [comment]
@mloskot, I've just prompted @hobu to start work on the first release candidate. If you can have a PR ready before he gets going I am with fine with removing both as of 5.0.0.

**Comment by mloskot** [comment]
@kbevers @hobu
* #754 removes Windows CE cruft
* #755 removes VB6 cruft

**Comment by kbevers** [comment]
@mloskot Thanks. Are you also familiar with the COM stuff too? I am wondering if it really is as easy as  `git rm -r com` to get rid of it.

**Comment by kbevers** [comment]
Oh, I see you got that covered as well!

**Comment by mloskot** [comment]
@kbevers 
> git rm -r com 

Yes, it was as easy as this for COM. I scanned the whole repository for "com" in relation to stuff in that folder, nothing found.

**Comment by kbevers** [comment]
Awesome, thanks. I'll merge once the various tests pass. Updating release notes now.

**Comment by kbevers** [comment]
I have created a bunch of [milestones](https://github.com/OSGeo/proj.4/milestones) which covers the essense of the proposed release schedule described in this issue. Closing this now. Once we have a final release of 5.0.0 I will start creating tickets for the various action items required for the various scheduled releases.
